### PR TITLE
avoid compiling boost code that is incompatible with C++17

### PR DESCRIPTION
**Pull request by boyceg**
### IBAMR Pull Request Checklist
- [ ] Does the test suite pass?
- [ ] Was clang-format run?
- [ ] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [ ] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [ ] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [ ] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [ ] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

File: ibtk/config/config.h.tmp [hunk 1]
```diff
@@ -66,4 +66,9 @@ _Pragma("GCC diagnostic ignored \"-Wunneeded-internal-declaration\"")
 #include <ibtk/compiler_hints.h>
 #include <ibtk/ibtk_utilities.h>
 
+// Tell boost to get with the program:
```
Comment (drwells) [inline ibtk/config/config.h.tmp:69]:
```suggestion
// Disable std::auto_ptr inside our bundled copy of boost:
```

**Review by drwells — CHANGES_REQUESTED** [review]

File: ibtk/config/config.h.tmp (hunk 1 above)
Comment (boyceg) [reply ibtk/config/config.h.tmp:69]: Unfortunately this macro does much more than just disable `auto_ptr`. We need it here to disable the use of `unary_function`.

File: ibtk/configure.ac [hunk 2]
```diff
@@ -94,7 +94,9 @@ dnl provided by argument)
 CONFIGURE_LIBMESH
 CONFIGURE_PETSC
 # Configure other libraries that are not themselves dependencies:
-CONFIGURE_BOOST("$ABSOLUTE_SRCDIR/contrib/boost","\$(abs_top_builddir)/contrib/boost")
+if test "$USING_BUNDLED_BOOST" = yes; then
+  CONFIGURE_BOOST("$ABSOLUTE_SRCDIR/contrib/boost","\$(abs_top_builddir)/contrib/boost")
+fi
```
Comment (drwells) [inline ibtk/configure.ac:99]: Is this correct? We already check for bundled boost in `configure_boost.m4`.

Comment (boyceg) [reply ibtk/configure.ac:99]: I am not sure, but I could not get `configure` to run without adding this check.

File: ibtk/config/config.h.tmp (hunk 1 above)
Comment (drwells) [reply ibtk/config/config.h.tmp:69]: How about `// Disable features which were deprecated in C++11 and removed in C++17 in boost:`

Comment (boyceg) [reply ibtk/config/config.h.tmp:69]: Isn't that what I wrote? 😁

Comment (drwells) [reply ibtk/config/config.h.tmp:69]: Technically but I didn't know that it was for the other deprecated then removed functions 🙈

Comment (boyceg) [reply ibtk/config/config.h.tmp:69]: I think this might be a MSVC flag that Boost has adopted...?

File: ibtk/configure.ac (hunk 2 above)
Comment (boyceg) [reply ibtk/configure.ac:99]: I will take another look when I get a few minutes.

**Comment by boyceg** [comment]
@drwells I've updated this PR. LMK if you have other comments. I am not sure the usual checklist applies.

**Review by drwells — APPROVED** [review]

**Comment by drwells** [comment]
Things pass the CI so this is good - thanks for taking a second look.

**Pull request merged**